### PR TITLE
Smart Contract Development – Major Refactor and Feature Expansion

### PR DESCRIPTION
**Pull request by kaneki003**
## Updates

- **Introduced Three New Auction Contracts**
  - Added Vickrey (sealed-bid),Exponential and Logarithmic Reverse Dutch auction contracts.
  - All auctions now support ERC20 token bidding, allowing for flexible payment options (including wrapped ETH if needed).
  - Enhanced logic for secure bidding, deadline management, and safe withdrawal handling.

- **Abstracted Common Auction Logic**
  - Created a new abstract `Auction` contract to centralize shared logic and enforce consistency across different auction types.
  - Standardized common events and auction type definitions.

- **Refactored Existing Codebase**
  - Migrated Linear Reverse Dutch auction to ERC20-based bidding.
  - Unified auction data structures, event naming conventions, and access control modifiers throughout the contracts.

- **Cleaned Up Legacy Code**
  - Removed old ERC20 all-pay and highest-bidder-pay contracts.

## Additional Notes

- All auctions now work with ERC20 tokens; ETH is supported through its wrapped form, based on team feedback and discussions.

---

## Summary by CodeRabbit

- **New Features**
  - Added a refactored AllPayAuction contract supporting ERC20 bidding tokens with explicit bid amounts and per-bidder tracking.
  - Introduced new auction contracts: EnglishAuction, VickreyAuction (sealed-bid), LinearReverseDutchAuction, ExponentialReverseDutchAuction, and LogarithmReverseDutchAuction, all supporting NFTs and ERC20 tokens.
  - Enabled flexible auction creation with detailed parameters including bidding tokens, deadlines, price decay models, and commit/reveal phases.
  - Implemented secure bidding, deadline extensions, dynamic price calculations, and robust withdrawal mechanisms for winners and auctioneers.

- **Bug Fixes**
  - Enhanced validation for auction parameters, bidding amounts, and withdrawal conditions to prevent invalid operations and multiple withdrawals.

- **Refactor**
  - Reorganized auction contracts to inherit from a common abstract Auction contract with unified event definitions and token transfer handling.
  - Replaced ETH-based bidding with ERC20 token transfers for clearer fund management.
  - Updated event names and parameters for consistency.

- **Revert**
  - Removed legacy auction contracts including AllPayAuctionERC20, HighestBidderPayAuction, and ReverseDutchAuction in favor of improved modular implementations.

- **Tests**
  - Added comprehensive test suites for new auction contracts covering creation, bidding, price decay, withdrawals, and edge cases.
  - Refactored AllPayAuction tests to support ERC20 bidding tokens and improved validation.
  - Removed tests for deprecated auction contracts.

### CODE REVIEW

**Comment by kaneki003** [comment]
@ceilican , @yogesh0509 ,I have done the contract development based on the feedback taken.Kindly review.
Also to make the review easier,i have squashed all the earlier commits into a single one to help you better track the changes.

File: contracts/AllPayAuction.sol [hunk 1]
```diff
@@ -1,193 +1,179 @@
...
 
-    enum AuctionType {
-        NFT,
-        Token
-    }
+    uint256 public auctionCounter=0;
+    mapping (uint256 => AuctionData) public auctions;
 
-    struct Auction {
+    struct AuctionData{
```
Comment (ceilican) [inline contracts/AllPayAuction.sol:18]: space missing.

File: contracts/AllPayAuction.sol [hunk 2]
```diff
@@ -1,193 +1,179 @@
...
 import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
-import "@openzeppelin/contracts/access/Ownable.sol";
 
-contract AllPayAuction is Ownable {
-    constructor() Ownable(msg.sender) {}
+/**
+ * @title AllPayAuction
+ * @notice Auction contract for NFT and token auctions,where all bidders pay their bid amount but only the highest bidder wins the auction.
+ */
+contract AllPayAuction is Auction{
```
Comment (ceilican) [inline contracts/AllPayAuction.sol:13]: space missing before "{".

File: contracts/EnglishAuction.sol [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+/**
+ * @title EnglishAuction
+ * @notice Auction contract for NFT and token auctions, where the highest bidder wins the auction and rest of the bidders get their bid refunded.
+ */
+contract EnglishAuction is Auction{
+
+    uint256 public auctionCounter=0;
+    mapping (uint256 => AuctionData) public auctions;
+
+    struct AuctionData{
```
Comment (ceilican) [inline contracts/EnglishAuction.sol:18]: Can't we move this struct and the `AuctionCreated` event to the `Auction` contract or somewhere else from where they could be reused instead of duplciated?

File: contracts/AllPayAuction.sol [hunk 4]
```diff
@@ -1,193 +1,179 @@
...
-            IERC721(auctionedTokenAddress).transferFrom(msg.sender,address(this),auctionedTokenIdOrAmount);
-        } else {
-            require(IERC20(auctionedTokenAddress).balanceOf(msg.sender) >= auctionedTokenIdOrAmount,"Caller must have sufficient tokens");
-            IERC20(auctionedTokenAddress).transferFrom(msg.sender,address(this),auctionedTokenIdOrAmount);
+        uint256 duration,
+        uint256 deadlineExtension
+    ) external{
+        require(bytes(name).length>0,"Name must be present");
+        require(minBidDelta>=0,"Minimum bid delta cannot be negative");
+        require(startingBid>=0,"Starting bid cannot be negative");
```
Comment (ceilican) [inline contracts/AllPayAuction.sol:74]: These "requires" are not necessary. Since `minBidDelta` and `startingBid` are uint256, they cannot be negative anyway.

File: contracts/AllPayAuction.sol [hunk 5]
```diff
@@ -1,193 +1,179 @@
...
+        require(duration>0,"Duration must be greater than zero seconds");
```
Comment (ceilican) [inline contracts/AllPayAuction.sol:75]: This is not necessary.

File: contracts/AllPayAuction.sol [hunk 6]
```diff
@@ -1,193 +1,179 @@
...
 
-        uint256 auctionId = auctionCounter;
-        auctions[auctionId] = Auction({
-            id: auctionCounter++,
-            name: name,
-            description: description,
-            imageUrl: imageUrl,
-            auctionType: auctionType,
-            itemWithdrawn: false,
+        uint256 deadline=block.timestamp+duration;
```
Comment (ceilican) [inline contracts/AllPayAuction.sol:88]: Generally, use spaces before and after operators such as `=`, `+`, `>=`, `==`.

File: contracts/EnglishAuction.sol [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+        uint256 deadline,
+        uint256 deadlineExtension
+    );
+
+    modifier validAuctionId(uint256 auctionId) {
+        require(auctionId>=0 && auctionId < auctionCounter,"Invalid auctionId");
+        _;
+    }
+
+    function createAuction(
```
Comment (ceilican) [inline contracts/EnglishAuction.sol:59]: I suspect that at least part of this function's code could be moved to the `Auction` contract as well, since it tends to be teh same for all auctions.

File: contracts/ExponentialReverseDutchAuction.sol [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        uint256 timeElapsed=block.timestamp-(auction.deadline-auction.duration);
+        uint256 x=timeElapsed * auction.decayFactor;
+        uint256 decayValue=getDecayValue(x);
+
+        // price(t) = reservedPrice + (startingPrice - reservedPrice) * 2^(-timeElapsed*decayFactor)
+        return auction.reservedPrice + ((auction.startingPrice-auction.reservedPrice)*decayValue)/1e18; 
+    }
+
+    //placeBid function is not required in this auction type as the price is determined by the auctioneer and the auctioneer can withdraw the item on placing the bid only
+    function withdrawItem(uint256 auctionId,uint256 bidAmount) external validAuctionId(auctionId) {
```
Comment (ceilican) [inline contracts/ExponentialReverseDutchAuction.sol:162]: We don't need `bidAmount` here, since the bidder will pay `currentPrice` anyway.

File: contracts/ExponentialReverseDutchAuction.sol [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+        AuctionData storage auction = auctions[auctionId];
+        require(block.timestamp<auction.deadline,"Auction has ended");
+        require(!auction.isClaimed,"Auction has been settled");
+
+        uint256 currentPrice=getCurrentPrice(auctionId);
+        require(bidAmount>=currentPrice,"Bid amount is less than current price");
+        
+        SafeERC20.safeTransferFrom(IERC20(auction.biddingToken),msg.sender,address(this),currentPrice);
+        if(auction.auctionType == AuctionType.NFT){
```
Comment (ceilican) [inline contracts/ExponentialReverseDutchAuction.sol:171]: I think we can create an auxiliary `transferItem` function in the `Auction` base contract that encapsulates the logic of checking whether the item is an NFT or amount of FT and calls the suitable transfer function.

File: contracts/ExponentialReverseDutchAuction.sol [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        AuctionData storage auction = auctions[auctionId]; 
+        require(msg.sender==auctions[auctionId].auctioneer,"Not auctioneer!");
+        uint256 withdrawAmount=auction.availableFunds;
+        require(withdrawAmount > 0,"No funds available");
+
+        SafeERC20.safeTransfer(IERC20(auction.biddingToken),auction.auctioneer,withdrawAmount);
+        auction.availableFunds=0;
+
+        emit fundsWithdrawn(
+            auctionId,
```
Comment (ceilican) [inline contracts/ExponentialReverseDutchAuction.sol:194]: There is no need to break lines here.  Look for other places where the formatter is exaggerating in the line breaks and eliminate the unnecessary line breaks.

File: contracts/LinearReverseDutchAuction.sol [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+        }
+
+        auction.winner=msg.sender;
+        auction.availableFunds=bidAmount;
+        auction.isClaimed=true;
+
+        emit itemWithdrawn(auctionId,msg.sender,auction.auctionedToken,auction.auctionedTokenIdOrAmount);
+    }
+
+    function withdrawFunds(uint256 auctionId) external validAuctionId(auctionId){
```
Comment (ceilican) [inline contracts/LinearReverseDutchAuction.sol:155]: I think the functions `withdrawFunds` from most auctions can be moved to the base `Auction` contract.

File: contracts/LinearReverseDutchAuction.sol [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+        AuctionData storage auction=auctions[auctionId];
+        require(block.timestamp<auction.deadline,"Auction has ended");
+        require(!auction.isClaimed,"Auction has ended");
+
+        // price(t) = startingPrice - (((startingPrice - reservedPrice) * (timeElapsed)) / duration)
+        return auction.startingPrice - (((auction.startingPrice-auction.reservedPrice)*(block.timestamp - (auction.deadline - auction.duration)))/auction.duration);
+    }
+
+    //placeBid function is not required in this auction type as the price is determined by the auctioneer and the auctioneer can withdraw the item on placing the bid only
+    function withdrawItem(uint256 auctionId,uint256 bidAmount) external validAuctionId(auctionId) {
```
Comment (ceilican) [inline contracts/LinearReverseDutchAuction.sol:133]: I think this function is the same for all ReverseDutchAuction types. Therefore, we can create an abstract superclass called `ReverseDutchAuction`, a subclass of `Auction`, containing this function. Then the subclasses of `ReverseDutchAuction` will just have to implement `getCurrentPrice`.

File: contracts/LogarithmReverseDutchAuction.sol [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+            biddingToken,
+            startingPrice,
+            reservedPrice,
+            decayFactor,
+            deadline
+        );
+    }
+
+    //Returns the logarithm base 2 of x scaled by 10^18 with a specified number of fractional bits.
+    function log2Fixed(uint256 x,uint8 fracBits) internal pure returns (uint256){
```
Comment (ceilican) [inline contracts/LogarithmReverseDutchAuction.sol:134]: Isn't there a library that we could use, instead of implementing our own logarithm functions?

File: contracts/VickreyAuction.sol [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+        );
+    }
+
+    function commitBid(uint256 auctionId,bytes32 commitment) validAuctionId(auctionId) external payable {
+        AuctionData storage auction = auctions[auctionId]; 
+        require(block.timestamp<auction.bidCommitEnd,"The commiting phase has ended!");
+        require(commitments[auctionId][msg.sender]==bytes32(0),"The sender has already commited");
+        require(msg.value > 1000000000000000, "Minimum fees not fulfilled");  //minimum fees of 0.001 ether
+        commitments[auctionId][msg.sender]=commitment; 
+    }
```
[the remark this answers is not in the document]
Comment (ceilican) [reply contracts/VickreyAuction.sol:130]: Correct. This is important.

**Review by ceilican — CHANGES_REQUESTED** [review]

File: contracts/AllPayAuction.sol (hunk 1 above)
Comment (yogesh0509) [reply contracts/AllPayAuction.sol:18]: We can implement a linter in our package.json along with a check on github pipelines to fix these in future easily.

File: contracts/EnglishAuction.sol (hunk 3 above)
Comment (kaneki003) [reply contracts/EnglishAuction.sol:18]: Yes @ceilican , i had thought of that but the problem was all the contracts aren't using same auction struct.For eg,reverse dutch contracts uses `decayFactor` ,`startingPrice` like terms ,which are not present in other auctions.In normal auctions also we have `minBidDelta` which is of no use to other auction contracts.

Since,`AuctionCreated` event is emitting auction params,with different params we couldn't store it common for all as well.

File: contracts/EnglishAuction.sol (hunk 7 above)
Comment (kaneki003) [reply contracts/EnglishAuction.sol:59]: I have moved modifiers and common requires of auction params while creating new auction into abstract contract along with `auctionCounter`  identifier

**Comment by kaneki003** [comment]
Changes made :

- Improved the abstraction of boilerplate code.
-  Added test suites to verify the functionality of the smart-contracts.
-  Implemented changes to prevent re-entrancy attacks, which helps improve the security of the smart contracts.
- Introduced a new formatter to enforce & improve code style consistency.

File: contracts/LogarithmReverseDutchAuction.sol [hunk 15]
```diff
@@ -1,22 +1,20 @@
...
  * The first bidder to meet the current price wins the auction.
+ * price(t) = startingPrice - ((startingPrice - reservedPrice) * log(1+k*t)) / log(1+k*duration)
  */
-contract LogarithmReverseDutchAuction is Auction{
-
-    uint256 public auctionCounter=0;
-    mapping (uint256 => AuctionData) public auctions;
-
-    struct AuctionData{
+contract LogarithmReverseDutchAuction is Auction {
```
Comment (ceilican) [inline contracts/LogarithmReverseDutchAuction.sol:15]: "Logarithm" => "Logarithmic", also in the file name.

**Review by ceilican — APPROVED** [review]
